### PR TITLE
ENH: Multithread PointSetToPointSetMetricv4

### PR DESCRIPTION
**Pull request by samuelgerber**
Add Multithreading to PointSetToPointSetMetricv4

Added multithreading support to itkPointSetToPointSetMetricv4.
This significantly improves the performance when used in conjunction with the performance fix in:
https://github.com/InsightSoftwareConsortium/ITK/pull/1037

The multithreading manually creates ranges to loop over the whole data.
Each loop sums and stores its total in a seperate variable order to avoid
false sharing with multiple threads accessing the same summation variable.

A text example can be found in:
https://github.com/InsightSoftwareConsortium/ITKTrimmedPointSetRegistration
that collects timings.
By setting ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS different numbers of threads can be tested.

## PR Checklist

- [x] :no_entry_sign: [Makes breaking changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes)
- [X] :no_entry_sign: [Makes design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes)
- [X] :no_entry_sign: Adds the License notice to new files.
- [X] :no_entry_sign: Adds Python wrapping.
- [X] :no_entry_sign: Adds tests and baseline comparison (quantitative).
- [X] :no_entry_sign: [Adds test data](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Documentation/UploadBinaryData.md).
- [X] :no_entry_sign: Adds Examples to [ITKExamples](https://github.com/InsightSoftwareConsortium/ITKExamples)
- [X] :no_entry_sign: Adds Documentation.

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

**Comment by ntustison** [comment]
Yes, this looks great.

**Comment by samuelgerber** [comment]
@ntustison If you have time I'd appreciate you checking and commenting on the changes I did to check if the transformed point sets need to be updated. In particular:

- I think the timestamp wasn't updated correctly which results in multiple reallocation of the point sets during each iteration (and associated construction of the KDTree). 
(Methods TransformFixedAndCreateVirtualPointSet() and TransformMovingPointSet() in PointSetToPointSetMetricv4 )

- I also added methods to check if the KDTree requires updating for a specific metric to avoid unnecessary constructions (e.g. the Euclidean metric only needs a KDTree on the moving point set which stays fixed and does not need to update the KdTree for the fixed transformed point set at every iteration).
(Methods RequiresFixedPointsLocator() and RequiresMovingPointsLocator() PointSetToPointSetMetricv4 and overrides in subclasses).

File: Modules/Registration/Metricsv4/include/itkPointSetToPointSetMetricv4.hxx [hunk 1]
```diff
@@ -287,7 +287,8 @@ PointSetToPointSetMetricv4<TFixedPointSet, TMovingPointSet, TInternalComputation
 
   PointIdentifierRanges ranges = this->CreateRanges();
```
Comment (dzenanz) [inline Modules/Registration/Metricsv4/include/itkPointSetToPointSetMetricv4.hxx]: Misspelling in commit message: derivtaive.

File: Modules/Registration/Metricsv4/test/itkObjectToObjectMultiMetricv4Test.cxx [hunk 2]
```diff
@@ -44,6 +44,10 @@ using ObjectToObjectMultiMetricv4TestMultiMetricType = itk::ObjectToObjectMultiM
 
 int itkObjectToObjectMultiMetricv4TestEvaluate( ObjectToObjectMultiMetricv4TestMultiMetricType::Pointer & multiVariateMetric, bool useDisplacementTransform )
 {
+  //The summations in the derivative depend on the ordering.
+  //Multithreading can introduce noticably different results
+  itk::MultiThreaderBase::New()->SetMaximumNumberOfThreads(1);
```
Comment (dzenanz) [inline Modules/Registration/Metricsv4/test/itkObjectToObjectMultiMetricv4Test.cxx]: Perhaps: `itk::MultiThreaderBase::SetGlobalMaximumNumberOfThreads(1);`?

Comment (samuelgerber) [reply Modules/Registration/Metricsv4/test/itkObjectToObjectMultiMetricv4Test.cxx]: I removed this change again. I added it due to the test initially failing which was fixed when the fix https://github.com/InsightSoftwareConsortium/ITK/pull/1037 was incorporated into this branch (the fix in 1037 removes the issue of threads writing/reading to the same temporary variable which messed up the results.)

**Pull request merged**